File: 07/main.py
import sys, re

def main():

    rules = dict()
    for rule in re.split("\n", sys.stdin.read().rstrip()):
        bags = dict()
        for bag_count in re.findall(" (\d+) (\w+ \w+) bags?[.,]", rule):
            bags[bag_count[1]] = int(bag_count[0])
        bag = re.match("(\w+ \w+) bags contain", rule)
        rules[bag.group(1)] = bags

    # find the total num of outer bags that can contain 1 shiny gold bag
    found1 = find_outer_bags(rules, 'shiny gold')
    print("PART 1:", len(found1))

    # find the total number of nested bags that are contained within 1 shiny gold bag
    found2 = find_nested_bags(rules, 'shiny gold')
    print("PART 2:", sum(map(lambda x: x[1] * x[2], found2)))

def find_outer_bags(rules, target, found=set()):
    for bag, contents in rules.items():
        if target in contents:
            found.add(bag)
            found = find_outer_bags(rules, bag, found)
    return found

def find_nested_bags(rules, target, count=1, found=list()):
    contents = rules[target]
    for bag_inside, bag_count in contents.items():
        found.append([bag_inside, count, bag_count])
        found = find_nested_bags(rules, bag_inside, count * bag_count, found)
    return found

if __name__ == '__main__':
    main()

File: 07/main.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
)

var bagRegexp = regexp.MustCompile(`^(\w+ \w+) bags contain`)
var contentsRegexp = regexp.MustCompile(` (\d+) (\w+ \w+) bags?[.,]`)

type Counter map[string]int
type Rules map[string]Counter

func main() {

	filePath := flag.String("file", "input.txt", "file containing the input data")
	flag.Parse()

	lines := mustReadFile(*filePath)
	rules := parseLines(lines)

	found1 := findOuterBags(rules, "shiny gold", make(Counter))
	fmt.Println("PART 1:", len(found1))

	found2 := findNestedBags(rules, "shiny gold", 1, make(Counter))
	answer2 := 0
	for _, count := range found2 {
		answer2 += count
	}
	fmt.Println("PART 2:", answer2)
}

func findOuterBags(rules Rules, target string, found Counter) Counter {
	for bag, contents := range rules {
		if _, ok := contents[target]; ok {
			found[bag] = 1
			found = findOuterBags(rules, bag, found)
		}
	}
	return found
}

func findNestedBags(rules Rules, target string, count int, found Counter) Counter {
	contents := rules[target]
	for bag, bagCount := range contents {
		found[bag] += (count * bagCount)
		found = findNestedBags(rules, bag, count*bagCount, found)
	}
	return found
}

func parseLines(lines []string) Rules {
	rules := make(Rules)
	for _, line := range lines {
		matches := bagRegexp.FindStringSubmatch(line)
		if matches == nil {
			log.Fatalf("cannot parse line: %s", line)
		}
		outerBag := mustParseString(matches[1])

		contentMatches := contentsRegexp.FindAllStringSubmatch(line, -1)
		for _, match := range contentMatches {
			quantity := mustParseInt(match[1])
			innerBag := mustParseString(match[2])
			if _, ok := rules[outerBag]; !ok {
				rules[outerBag] = make(Counter)
			}
			rules[outerBag][innerBag] = quantity
		}
	}
	return rules
}

func mustReadFile(path string) []string {
	if path == "" {
		log.Fatalf("file path not specified: %v", path)
	}

	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("cannot open file: %v", err)
	}
	defer file.Close()

	lines := make([]string, 0)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("scanner error: %v", err)
	}

	return lines
}

func mustParseInt(s string) int {
	i, err := strconv.Atoi(s)
	if err != nil {
		log.Fatalf("cannot convert string %s to integer: %v", s, err)
	}
	return i
}
func mustParseString(s string) string {
	if s == "" {
		log.Fatalf("string cannot be empty")
	}
	return s
}
